### PR TITLE
Better DeepAR lags for business day frequency time series.

### PR DESCRIPTION
**Pull request by sighellan**
*Issue #, if available:*

*Description of changes:*

Currently the default lags for the DeepAR model on business day frequency data is just [1, 2, 3, 4, 5, 6, 7]. This pull request implements a version of the lags for daily data, but assuming 5 days in a week and 22 days in a month.

It ignores that not all Mondays-Fridays are business days.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

**Comment by sighellan** [comment]
Tested on a business day version of the weather dataset:

```python
from gluonts.mx.model.deepar import DeepAREstimator
from gluonts.mx import Trainer
from gluonts.evaluation import Evaluator
from gluonts.evaluation import make_evaluation_predictions
from gluonts.mx.trainer.callback import TrainingHistory
from gluonts.mx.trainer.callback import Callback
import logging
import numpy as np
import mxnet as mx
import datetime
import pandas
from gluonts.time_feature.lag import get_lags_for_frequency
from gluonts.dataset.repository.datasets import get_dataset
import datetime
logger = logging.getLogger(__name__)

class GluonTSEarlyStoppingCallback(Callback):
    # From https://github.com/awslabs/autogluon/blob/v0.5.2/timeseries/src/autogluon/timeseries/models/gluonts/callback.py

def convert_to_business_days(ts):
    bs_start = None
    ii = 0
    while bs_start is None:
        delta = datetime.timedelta(days=ii)
        if (ts['start']+delta).weekday not in [5, 6]:
            bs_start = pandas.Period(value=ts['start']+delta, freq='B')
        ii += 1
        
    business_series = []
    offset = ts['start'].weekday
    for ii in range(len(ts['target'])):
        if (ii+offset) % 7 not in [5, 6]:
            business_series.append(ts['target'][ii])
    bs_dict = {
        'target': np.array(business_series),
        'start': bs_start,
        'item_id': ts['item_id']
    }
    return bs_dict

dataset = get_dataset('weather')
business_pred_length = 22

train_list = list(dataset.train)
test_list = list(dataset.test)

print(len(train_list), len(train_list[0]['target']))
print(len(test_list), len(test_list[0]['target']))

business_train_list = [convert_to_business_days(ts) for ts in train_list]
business_test_list = [convert_to_business_days(ts) for ts in test_list]

print(len(business_train_list), len(business_train_list[0]['target']))
print(len(business_test_list), len(business_test_list[0]['target']))
```

Output: 
```
3010 4225
3010 4255

3010 3017
3010 3039

```

Test runs:
```python
epochs = 40
patience = 10

seeds = [1, 2, 3, 4, 5]
lag_lengths = [7, None]

freq = 'B'
full_lags = get_lags_for_frequency(freq)
pred_length = business_pred_length

res_dict = {}
for seed in seeds:

    for lag_len in lag_lengths:

        lags = full_lags[:lag_len]
        
        key = str(lags)
        print(key, seed)
            
        history = TrainingHistory()
        early_stopping = GluonTSEarlyStoppingCallback(patience=patience)

        trainer = Trainer(epochs=epochs, callbacks=[early_stopping, history])
        np.random.seed(seed)
        mx.random.seed(seed, trainer.ctx)

        estimator = DeepAREstimator(
                    prediction_length=pred_length, 
                    freq=freq, 
                    trainer=trainer,
                    lags_seq=lags
        )

        model = estimator.train(training_data=business_train_list[:-1],
                               validation_data=[business_train_list[-1]])

        forecast_it, ts_it = make_evaluation_predictions(
            dataset=business_test_list,  # test dataset
            predictor=model,  # predictor
            num_samples=100,  # number of sample paths we want for evaluation
        )

        forecasts = list(forecast_it)
        tss = list(ts_it)

        evaluator = Evaluator(quantiles=[0.1, 0.5, 0.9])
        agg_metrics, item_metrics = evaluator(tss, forecasts)

        print(estimator.lags_seq, agg_metrics['mean_wQuantileLoss'])
                    
        if key in res_dict:
            res_dict[key].append(agg_metrics)
        else:
            res_dict[key] = [agg_metrics]

def print_rd(num):
    print(np.round(num, 4))
for key in res_dict:
    print(key)
    loss = [agg['mean_wQuantileLoss'] for agg in res_dict[key]]
    print_rd(loss)
    mean, std = np.mean(loss), np.std(loss, ddof=1)/np.sqrt(len(loss))
    print_rd([mean-std, mean+std])
``` 

Output:
```
[1, 2, 3, 4, 5, 6, 7]
[0.1895 0.142  0.1578 0.1435 0.1866]
[0.1536 0.1742]
[1, 2, 3, 4, 5, 6, 7, 9, 10, 11, 14, 15, 16, 19, 20, 21, 22, 23, 39, 59, 258, 259, 260, 518, 519, 520, 778, 779, 780]
[0.1618 0.1261 0.1437 0.1402 0.1338]
[0.1351 0.1471]

```

So 'mean_wQuantileLoss' +- std of mean improves from 0.1536 - 0.1742 to 0.1351 - 0.1471.

**Comment by jaheba** [comment]
Thanks! Just a minor comment: We also have a time-limit callback now.

https://github.com/awslabs/gluonts/blob/0e4a488946fbdd8bf5a4da73997f399b575fda50/src/gluonts/mx/trainer/callback.py#L381-L389

**Comment by sighellan** [comment]
> Thanks! Just a minor comment: We also have a time-limit callback now.
> 
> https://github.com/awslabs/gluonts/blob/0e4a488946fbdd8bf5a4da73997f399b575fda50/src/gluonts/mx/trainer/callback.py#L381-L389

Cool! The early stopper is slightly different in that it looks at how the model performance changes with time to decide whether to halt, but this is useful as well!

**Review by lostella — APPROVED** [review]
